universe/supplyverifier/manager.go

```diff
@@ -0,0 +1,567 @@
+// new one if it doesn't exist. If a new state machine is created, it is also
+// started.
+func (m *Manager) fetchStateMachine(assetSpec asset.Specifier) (*StateMachine,
+	error) {
+
+	groupKey, err := assetSpec.UnwrapGroupKeyOrErr()
+	if err != nil {
+		return nil, fmt.Errorf("asset specifier missing group key: %w",
+			err)
+	}
+
+	// Check if the state machine for the asset group already exists in the
+	// cache.
+	sm, ok := m.smCache.Get(*groupKey)
+	if ok {
+		return sm, nil
+	}
+
+	// If the state machine is not found, create a new one.
+	env := &Environment{
+		AssetSpec:        assetSpec,
+		Chain:            m.cfg.Chain,
+		SupplyCommitView: m.cfg.SupplyCommitView,
+		ErrChan:          m.cfg.ErrChan,
+		QuitChan:         m.Quit,
+	}
+
+	// Before we start the state machine, we'll need to fetch the current
+	// state from disk, to see if we need to emit any new events.
+	ctx, cancel := m.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	initialState, err := m.cfg.StateLog.FetchState(ctx, assetSpec)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch current state: %w", err)
+	}
+
+	// Create a new error reporter for the state machine.
+	errorReporter := NewErrorReporter(assetSpec)
+
+	fsmCfg := protofsm.StateMachineCfg[Event, *Environment]{
+		ErrorReporter: &errorReporter,
+		InitialState:  initialState,
+		Env:           env,
+		Daemon:        m.cfg.DaemonAdapters,
+	}
+	newSm := protofsm.NewStateMachine[Event, *Environment](fsmCfg)
+
+	// Ensure that the state machine is running. We use the manager's
+	// context guard to derive a sub context which will be cancelled when
+	// the manager is stopped.
+	smCtx, _ := m.WithCtxQuitNoTimeout()
+	newSm.Start(smCtx)
+
+	// For supply verifier, we always start with an InitEvent to begin
+	// the verification process.
+	newSm.SendEvent(ctx, &InitEvent{})
+
+	m.smCache.Set(*groupKey, &newSm)
+
+	return &newSm, nil
+}
+
+// VerifySupplyLeaves is used to verify supply leaves for a given asset
+func (m *Manager) VerifySupplyLeaves(ctx context.Context,
+	assetSpec asset.Specifier, leaves supplycommit.SupplyLeaves) error {
+
+	_, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	// TODO(ffranr): Forward the supply leaves to the state machine.
+
+	return nil
+}
+
+// InsertSupplyCommit stores a verified supply commitment for the given asset
+// group in the node's local database.
+func (m *Manager) InsertSupplyCommit(ctx context.Context,
+	assetSpec asset.Specifier, commitment supplycommit.RootCommitment,
+	leaves supplycommit.SupplyLeaves) error {
+
+	headerVerifier := tapgarden.GenHeaderVerifier(ctx, m.cfg.Chain)
+
+	// Do the static on-chain verification of the supply commitment.
+	err := commitment.Verify(proof.DefaultMerkleVerifier, headerVerifier)
+	if err != nil {
+		return fmt.Errorf("unable to verify supply commitment: %w", err)
+	}
+
```

Comment:
Should also verify delegation key information, otherwise impersonation is possible.